src/balloonpanel/balloonpanelview.js

```diff
@@ -0,0 +1,371 @@
+		const panelRect = new AbsoluteDomRect( this.element );
+		const visibleInViewportRect = getVisibleInViewportRect( limiter );
+
+		// Get bounding box of each available place and move balloon in this place where fits the best.
+		this._smartAttachTo( [
+			// South east.
+			panelRect.clone( 'se' ).moveTo( {
+				top: elementOrRangeRect.bottom + arrowTopOffset,
+				left: elementOrRangeRect.left + elementOrRangeRect.width / 2 - arrowLeftOffset
+			} ),
+
+			// South west.
+			panelRect.clone( 'sw' ).moveTo( {
+				top: elementOrRangeRect.bottom + arrowTopOffset,
+				left: elementOrRangeRect.left + elementOrRangeRect.width / 2 - panelRect.width + arrowLeftOffset
+			} ),
+
+			// North east.
+			panelRect.clone( 'ne' ).moveTo( {
+				top: elementOrRangeRect.top - panelRect.height - arrowTopOffset,
+				left: elementOrRangeRect.left + elementOrRangeRect.width / 2 - arrowLeftOffset
+			} ),
+
+			// North west.
+			panelRect.clone( 'nw' ).moveTo( {
+				top: elementOrRangeRect.top - panelRect.height - arrowTopOffset,
+				left: elementOrRangeRect.left + elementOrRangeRect.width / 2 - panelRect.width + arrowLeftOffset
+			} )
+		], visibleInViewportRect );
+	}
+
+	/**
+	 * Move balloon in the place where fits the best.
+	 *
+	 * @private
+	 * @param {Array<{AbsoluteDomRect}>} rects List of positions where balloon can be placed.
+	 * @param {AbsoluteDomRect} visibleViewportRect Bounding box of visible part of viewport element.
+	 */
+	_smartAttachTo( rects, visibleViewportRect ) {
+		let maxIntersectRect;
+		let maxIntersectArea = -1;
+
+		// Get best place.
+		for ( let rect of rects ) {
+			const intersectArea = rect.getIntersectArea( visibleViewportRect );
+
+			if ( intersectArea > maxIntersectArea ) {
+				maxIntersectRect = rect;
+				maxIntersectArea = intersectArea;
+			}
+		}
+
+		// Move balloon.
+		this.model.arrow = maxIntersectRect.name;
+		this.model.top = maxIntersectRect.top;
+		this.model.left = maxIntersectRect.left;
+	}
+}
+
+/**
+ * The balloon panel view {@link ui.Model model} interface.
+ *
+ * @interface ui.balloonPanel.BalloonPanelViewModel
+ */
+
+/**
+ * Top position of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#top
+ */
+
+/**
+ * Left position of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#left
+ */
+
+/**
+ * Max width of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#maxWidth
+ */
+
+/**
+ * Balloon panel arrow direction.
+ *
+ * @observable
+ * @member {'se'|sw'|'ne'|nw'} ui.balloonPanel.BalloonPanelViewModel#arrow
+ */
+
+/**
+ * Controls whether the balloon panel is visible or not.
+ *
+ * @observable
+ * @member {Boolean} ui.balloonPanel.BalloonPanelViewModel#isVisible
+ */
+
+/**
+ * Private helper for count and manipulate bounding box of passed element relative to whole document.
```

Comment:
There's no counting done there. At least not in such way the verb would suggest.

"An abstract class which represents a bounding box of an HTMLElement or a Range in DOM."

 That's it.